cpp/source/deal.II/new_supercapacitor.templates.h

```diff
@@ -0,0 +1,358 @@
+      constant_current_density;
+  evolve_one_time_step(time_step, ConstantCurrent, rebuild);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_voltage(
+    double const time_step, double const voltage)
+{
+  bool const rebuild =
+      (electrochemical_physics_params->constant_voltage == voltage) ? false
+                                                                    : true;
+  electrochemical_physics_params->constant_voltage = voltage;
+  evolve_one_time_step(time_step, ConstantVoltage, rebuild);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_power(
+    double const time_step, double const power)
+{
+  BOOST_ASSERT_MSG(surface_area != 0., "The surface area is zero.");
+  dealii::Vector<double> old_solution(solution->block(0));
+  int const max_iterations       = 10;
+  double const percent_tolerance = 1.0e-2;
+  double current(0.0);
+  double voltage(0.0);
+  get_voltage(voltage);
+  for (int k = 0; k < max_iterations; ++k)
+  {
+    current                               = power / voltage;
+    double const constant_current_density = current / surface_area;
+    bool const rebuild =
+        (electrochemical_physics_params->constant_current_density ==
+         constant_current_density)
+            ? false
+            : true;
+    electrochemical_physics_params->constant_current_density =
+        constant_current_density;
+    evolve_one_time_step(time_step, ConstantCurrent, rebuild);
+    get_voltage(voltage);
+    if (std::abs(power - voltage * current) / std::abs(power) <
+        percent_tolerance)
+      return;
+    solution->block(0) = old_solution;
+  }
+  throw std::runtime_error("fixed point iteration did not converge in " +
+                           std::to_string(max_iterations) + " iterations");
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_load(
+    double const time_step, double const load)
+{
+  electrochemical_physics_params->constant_load_density = load * surface_area;
+  // BC not implemented yet
+  throw std::runtime_error("This function is not implemented.");
+  evolve_one_time_step(time_step, ConstantLoad, true);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_linear_current(
+    double const time_step, double const current)
+{
+  // TODO: this is a temporary solution
+  evolve_one_time_step_constant_current(time_step, current);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_linear_voltage(
+    double const time_step, double const voltage)
+{
+  // TODO: this is a temporary solution
+  evolve_one_time_step_constant_voltage(time_step, voltage);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_linear_power(
+    double const time_step, double const power)
+{
+  std::ignore = time_step;
+  std::ignore = power;
+
+  throw std::runtime_error("This function is not implemented");
```

Comment:
I'd map all of them to the ``constant`` version and add the ``TODO`` comment line